app/models/product.rb

```diff
@@ -71,17 +71,50 @@ def measure_ids
 
   def add_product_tests_to_product(added_measure_ids = [])
     return if added_measure_ids.nil?
+
     new_ids = added_measure_ids - measure_ids
     to_remove_ids = measure_ids - added_measure_ids
-
-    new_ids.each do |new_measure_id|
-      measure = Measure.top_level.find_by(hqmf_id: new_measure_id)
-      product_tests.build({ name: measure.name, product: self, measure_ids: [new_measure_id],
-                            cms_id: measure.cms_id, bundle_id: measure.bundle_id }, MeasureTest)
+    untouched_ids = measure_ids - to_remove_ids
+    if c1_test || c2_test || c3_test
+      new_ids.each do |new_measure_id|
+        measure = Measure.top_level.find_by(hqmf_id: new_measure_id)
+        product_tests.build({ name: measure.name, product: self, measure_ids: [new_measure_id],
+                              cms_id: measure.cms_id, bundle_id: measure.bundle_id }, MeasureTest)
+      end
     end
 
-    to_remove_ids.each do |old_measure_id|
-      product_tests.in(measure_ids: old_measure_id).destroy
+    to_remove_ids.each { |old_measure_id| product_tests.in(measure_ids: old_measure_id).destroy }
+
+    add_filtering_tests(ApplicationController.helpers.pick_measure_for_filtering_test(untouched_ids + new_ids)) if c4_test
+  end
+
+  def add_filtering_tests(measure)
+    save
+    reload_relations
+    filtering_tests = product_tests.select { |product_test| product_test.is_a? FilteringTest }
```

Comment:
We can probably trim this down a bit by rewriting this to just count the number of filter tests.  product_tests.count(_type: "FilterTest") .  Could/should create scoped query on product to avoid selecting and filtering test types .